ui/src/app/utils/subscribable.tsx

```diff
@@ -0,0 +1,58 @@
+import * as fp from 'lodash/fp';
+
+/**
+ * @name Subscribable
+ * @description A container allowing components and functions to subscribe to changes in the container's contents
+ * @function subscribe Add a subscription by passing a function
+ * @function subscribe Send the new value to all subscribers
+ **/
+export interface Subscribable<T> {
+  subscribe: (fn: (newValue?: T, oldValue?: T) => void) => { unsubscribe: () => void};
+  next: (newValue: T, oldValue: T) => void;
+}
+
+/** @name Atom
+ * @description A container that can store a value to be accessed and subscribed to globally
+ * @function get Returns the current value
+ * @function set Sets the value and triggers all subscribed functions with the old/new values
+ * @function subscribe Add a function to the subscribe list
+ * @function reset Resets the value to the initial value
+ */
+export interface Atom<T> {
+  get: () => T;
+  set: (value: T) => void;
+  subscribe: Subscribable<T>['subscribe'];
+  reset: () => void;
+}
+
+export function subscribable<T>(): Subscribable<T> {
+  let subscribers = [];
+  type subscriber = ((newValue?: T, oldValue?: T) => void);
+
+  return {
+    subscribe: fn => {
+      subscribers = fp.concat(subscribers, [fn]);
+      return {
+        unsubscribe: () => {
+          subscribers = fp.without([fn], subscribers);
+        }
+      };
+    },
+    next: (newValue?: T, oldValue?: T) => {
+      fp.forEach((fn: subscriber) => setTimeout(() => fn(newValue, oldValue), 0), subscribers);
+    }
+  };
+}
+
+// A simple state container inspired by clojure atoms.
+export function atom<T>(initialValue: T): Atom<T> {
```

Comment:
I googled this pattern cause I hadn't heard it called 'atom' before. It looks like this is a clojurescript pattern, and that [someone ported it](https://github.com/derrickbeining/react-atom) to react. Do you think this port is worth using, or can our needs be served by something as simple as what you've implemented here?